sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -205,6 +224,108 @@ case class BroadcastNestedLoopJoinExec(
+              s" fallback to leftExistenceJoin.")
+            false
+          }
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  private def notInSubquerySingleColumnOptimizeBuildParams(
+      buildRows: Array[InternalRow]): NotInSubquerySingleColumnOptimizeParams = {
+    NotInSubquerySingleColumnOptimizeParams(
+      HashedRelation(buildRows.iterator,
+        BindReferences.bindReferences[Expression](
+          Seq(right.output.head), AttributeSeq(right.output)),
+        buildRows.length),
+      buildRows.exists(row => row.isNullAt(0)),
+      buildRows.isEmpty)
+  }
+
+  private def notInSubquerySingleColumnOptimizeSetStreamedKey(
+      leftAttr: AttributeReference,
+      rightAttr: AttributeReference): Unit = {
+    val leftNotInKeyFound = left.output.exists(_.semanticEquals(leftAttr))
+    if (leftNotInKeyFound) {
+      notInSubquerySingleColumnOptimizeStreamedKey = leftAttr
+      notInSubquerySingleColumnOptimizeStreamedKeyIndex =
+        AttributeSeq(left.output).indexOf(leftAttr.exprId)
+    } else {
+      notInSubquerySingleColumnOptimizeStreamedKey = rightAttr
+      notInSubquerySingleColumnOptimizeStreamedKeyIndex =
+        AttributeSeq(left.output).indexOf(rightAttr.exprId)
+    }
+  }
+
+  /**
+   * Optimized version implementation for LeftAnti
+   * which converted from single column NotInSubquery
+   * @param relation
+   * @return
+   */
+  private def notInSubquerySingleColumnOptimize(
+      relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+    assert(buildSide == BuildRight)
+    val buildRows = relation.value
+    if (buildRows.length > SQLConf.get.notInSubquerySingleColumnOptimizeRowCountThreshold) {
+      logWarning(s"BuildSide row count ${buildRows.length} exceeded " +
+        s"notInSubquerySingleColumnOptimizeRowCountThreshold ${SQLConf.get
+          .notInSubquerySingleColumnOptimizeRowCountThreshold}, fallback to leftExistenceJoin.")
+      leftExistenceJoin(relation, false)
+    } else {
+      val params = notInSubquerySingleColumnOptimizeBuildParams(buildRows)
+      streamed.execute().mapPartitionsInternal { streamedIter =>
+        val keyGenerator =
+          UnsafeProjection.create(
+            BindReferences.bindReferences[Expression](
+              Seq(notInSubquerySingleColumnOptimizeStreamedKey),
+              AttributeSeq(left.output))
+          )
+        streamedIter.filter(row => {
+          // See. not-in-unit-tests-single-column.sql for detail filter rules
+          if (params.isBuildRowsEmpty) {
+            true
+          } else {
+            val streamedRowIsNull = row.isNullAt(notInSubquerySingleColumnOptimizeStreamedKeyIndex)
+            val lookupRow: UnsafeRow = keyGenerator(row)
+            val notInKeyEqual = params.buildSideHashedRelation.get(lookupRow) match {
+              case null => false
+              case _ => true
+            }
+
+            if (!streamedRowIsNull && !params.isNullExists && !notInKeyEqual) {
```

Comment:
The check for isNullExists can also be pulled out: If isNullExists, then we will unconditionally return nothing.

You have already paid the one time cost of doing a scan on the build size to check if any nulls exist (when preparing the params), you might as well exploit that check to gain some speed here :-)

Reply:
done.